servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSImpl.java

```diff
@@ -10098,6 +10137,144 @@ public void deleteAssertionCondition(ResourceContext ctx, String domainName, Str
         dbService.executeDeleteAssertionCondition(ctx, domainName, policyName, assertionId, conditionId, auditRef, caller);
     }
 
+    @Override
+    public void putDomainDependency(ResourceContext ctx, String domainName, String auditRef, String service) {
+        final String caller = ctx.getApiName();
+        logPrincipal(ctx);
+
+        if (readOnlyMode.get()) {
+            throw ZMSUtils.requestError(SERVER_READ_ONLY_MESSAGE, caller);
+        }
+
+        // If system administrator - service is mandatory
+        // Otherwise ignore it and use the principal as dependent service
+
+        Principal principal = ((RsrcCtxWrapper) ctx).principal();
+        boolean isSysAdminUser = isSysAdminUser(principal);
+        if (isSysAdminUser) {
+            if (StringUtil.isEmpty(service)) {
+                throw ZMSUtils.requestError("service is mandatory", caller);
+            }
+        } else {
+            service = principal.getFullName();
+        }
+
+        validateRequest(ctx.request(), caller);
+
+        validate(domainName, TYPE_DOMAIN_NAME, caller);
+        validate(service, TYPE_SERVICE_NAME, caller);
+
+        // for consistent handling of all requests, we're going to convert
+        // all incoming object values into lower case (e.g. domain, role,
+        // policy, service, etc name)
+
+        domainName = domainName.toLowerCase();
+        setRequestDomain(ctx, domainName);
+        service = service.toLowerCase();
+
+        // verify that service is a service provider
+
+        if (!serviceProviderManager.isServiceProvider(service)) {
+            throw ZMSUtils.unauthorizedError("Service " + service + " is not an authorized service provider", caller);
+        }
+
+        // Verify the domain exist
+
+        AthenzDomain domain = getAthenzDomain(domainName, false);
+        if (domain == null || domain.getDomain() == null) {
+            throw ZMSUtils.notFoundError("Invalid domain: " + domainName, caller);
+        }
+
+        dbService.putDomainDependency(ctx, domainName, service, auditRef, caller);
+    }
+
+    @Override
+    public void deleteDomainDependency(ResourceContext ctx, String domainName, String service, String auditRef) {
+        final String caller = ctx.getApiName();
+        logPrincipal(ctx);
+
+        if (readOnlyMode.get()) {
+            throw ZMSUtils.requestError(SERVER_READ_ONLY_MESSAGE, caller);
+        }
+
+        // If system administrator - service is mandatory
+        // Otherwise ignore it and use the principal as dependent service
+
+        Principal principal = ((RsrcCtxWrapper) ctx).principal();
```

Comment:
this block is the same as in the putDomainDependency call, so let's move into a local private method and would be simpler to make sure it's fully tested with all the if/else statements there.

Reply:
Fixed